### DISCUSSION TITLE
plantuml support out of box like mermaid

### DISCUSSION THREAD

**Discussion by plmcoder** [quarto-dev/quarto-cli#6549] [Q&A] [answered]
### Description

Hi,
plantuml is my favorite UML tool than mermaid, looks like more professional than mermaid. and lots of programmers use it. 

would you please support plantuml out of box in next version of Quarto?
I tried for weeks but could not find an extension to integrate this into quarto.

- `iplantuml` not work.
- `quarto-ext/diagram` not work.
- `python plantuml-markdown` not work.
- ...

all are not work.

**Comment by plmcoder** [comment]
It not easy to me to learn pandoc api + lua script language  to write a new one.
Lua is not very difficult, but we have to spend lots of time to learn pandoc api and latex.
so I ask you for help.

thanks a lot!

**Comment by mcanouil** [comment]
Did you look at:

- #819

Which highlight a Pandoc filter: https://github.com/pandoc/lua-filters/tree/master/diagram-generator#plantuml which was ported a a Quarto extension, see <https://github.com/pandoc-ext/diagram>.

Also "not work" is not really something that help us help you.
If you could share code, Git repository, etc. in which it works or not as expected would help others help you.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by plmcoder** [reply]
thanks for your reminder!

**Comment by plmcoder** [comment]
Hi Mcanouil,

follow your help it works now! thanks a lot for your help.
Here is my steps:
1. install anaconda 23.7.2 (with python3)
2. conda install -c conda-forge quarto
3. download TinyTeX from github.com and unzip TinyTeX to D:\apps\TinyTeX
4. set PLANTUML= D:\apps\plantuml.jar (not work), place plantuml.jar in workdir instead.
5. other enviroments:
set JAVA_HOME=D:\apps\jdk-17.0.8+7
set DOT=D:\apps\Graphviz
set TEX_HOME=D:\apps\TinyTeX

set PATH=%PATH%;%JAVA_HOME%\bin;%DOT%\bin;%TEX_HOME%\bin;


6.plantuml：
    6.1 conda install -c conda-forge librsvg # missing this will go error while generate pdf format.
    6.2 copy plantuml.jar to workdir.
    6.3 copy diagram-generator.lua to workdir.
    6.4 set JAVA_HOME path.
    6.5 create test-diagram-generator.qmd and run quarto render test-diagram-generator.qmd



test-diagram-generator.qmd file:
> please replace `'''` to ` ``` `.
> must keep ` keep-tex: true` because mediabag folder will be deleted and the html format will got error.

```
---
title: xxx
subtitle: xxx.
date: today
lang: zh
author:
  - name: Alex XIAO
execute:
  echo: false
format:
  docx: 
    number-sections: true
    toc: true
    lof: true
    lot: true
    keep-tex: true
  pptx:
    number-sections: true
    toc: true
    lof: true
    lot: true
    keep-tex: true
  pdf: 
    number-sections: true
    toc: true
    lof: true
    lot: true
    keep-tex: true
  html: 
    mermaid:
      theme: neutral
    grid:
      sidebar-width: 300px
      body-width: 1200px
      margin-width: 300px
      gutter-width: 1.5rem
    number-sections: true
    toc: true
    lof: true
    lot: true
filters:
  - diagram-generator.lua
---

# test mermaid

'''{mermaid}
sequenceDiagram
    participant Alice
    participant Bob
    Alice->>John: Hello John, how are you?
    loop Healthcheck
        John->>John: Fight against hypochondria
    end
    Note right of John: Rational thoughts <br/>prevail!
    John-->>Alice: Great!
    John->>Bob: How about you?
    Bob-->>John: Jolly good!
'''

# test plantuml

'''{.plantuml caption="xxxxxx" width=50%}
@startuml
entity "SimTask(仿真科目)" as simtask{
}

class "Param(性能指标)" as param{
}
simtask "1" o-- "N" param:TaskParams

@enduml

'''

'''{.plantuml caption="yyyyy" width=50%}
@startuml
skinparam Shadowing false
Alice -> Bob: Authentication Request
Bob -> RequestHandler: validated Authenticated Request
RequestHandler --> Bob: Authenticated Response (OK)
Bob --> Alice: Authentication Response (OK)

@enduml
'''

```

**Reply by mcanouil** [reply]
Could you use the syntax I showed in my post instead of asking to replace characters?
Also why are the defining for each format the same things? Simply put it at the top level to apply the options to all formats.

```
---
title: xxx
subtitle: xxx.
date: today
lang: zh
author:
  - name: Alex XIAO
execute:
  echo: false
number-sections: true
toc: true
lof: true
lot: true
keep-tex: true
format:
  docx: default
  pptx: default
  pdf: default
  html: 
    mermaid:
      theme: neutral
    grid:
      sidebar-width: 300px
      body-width: 1200px
      margin-width: 300px
      gutter-width: 1.5rem
filters:
  - diagram-generator.lua
---
```

**Comment by cderv** [comment, marked as answer]
Plantuml support lives in https://github.com/pandoc-ext/diagram 

If something is not working, this should be opened there.
